pypeit/core/flat.py

```diff
@@ -212,88 +238,136 @@ def fit_flat(flat, mstilts, thismask, slit_left, slit_righ, inmask = None,spec_s
-    else:
-        lmed = normimg[0]
-        rmed = normimg[-1]
+    # Determine the maximum at the left and right end of the slit
+    ileft = (ximg_fit[imed] > 0.1) & (ximg_fit[imed] < 0.4)
+    xleft = ximg_fit[imed][ileft]
+    norm_max_left = normimg[ileft].max()
+    xmax_left = xleft[normimg[ileft].argmax()]
+    irigh = (ximg_fit[imed] > 0.6) & (ximg_fit[imed] < 0.9)
+    xrigh = ximg_fit[imed][irigh]
+    norm_max_righ = normimg[irigh].max()
+    xmax_righ = xrigh[normimg[irigh].argmax()]
+
+
     # mask regions where illumination function takes on extreme values
     if np.any(~np.isfinite(normimg)):
         msgs.error('Inifinities in slit illumination function computation normimg')
-    illum_max = (np.abs(normimg[statinds]/mean - 1.0)).max()
-    if (illum_max <= spat_illum_thresh):
-        ximg_in = np.concatenate((-0.2 + 0.2*np.arange(npad)/(npad - 1), ximg_fit,1.0 + 0.2*np.arange(npad)/(npad-1)))
-        normin = np.ones(2*npad + nfit_spat)
-        #msgs.info('illum_max={:7.3f}'.format(illum_max))
-        msgs.info('Illum fluctuations = {:7.3f}'.format(illum_max*100) +
-                  '% < spat_illum_thresh={:4.2f}'.format(100.0*spat_illum_thresh)+'%')
-        msgs.info('Slit illumination function set to unity for this slit%')
-        no_illum = True
-    else:
-        ximg_in = np.concatenate((-0.2 + 0.2*np.arange(npad)/(npad-1), ximg_fit[imed], 1.0 + 0.2*np.arange(npad)/(npad-1)))
-        normin =  np.concatenate((np.full(npad, lmed), normimg, np.full(npad,rmed)))
-
-    #    if spat_bkpt is not None:
-    #        fullbkpt = spatbkpt
-    #    else:
 
     # Determine the breakpoint spacing from the sampling of the ximg
     ximg_samp = np.median(ximg_fit - np.roll(ximg_fit,1))
     ximg_1pix = 1.0/slitwidth
     # Use breakpoints at a spacing of a 1/10th of a pixel, but do not allow a bsp smaller than the typical sampling
     ximg_bsp  = np.fmax(ximg_1pix/10.0, ximg_samp*1.2)
-    bsp_set = pydl.bspline(ximg_in,nord=4, bkspace=ximg_bsp)
+    bsp_set = pydl.bspline(ximg_fit,nord=4, bkspace=ximg_bsp)
     fullbkpt = bsp_set.breakpoints
-    spat_set, outmask_spat, spatfit, _ = utils.bspline_profile(ximg_in, normin, np.ones_like(normin),np.ones_like(normin),
-                                                               nord=4,upper=5.0, lower=5.0,fullbkpt = fullbkpt)
+    normfit = np.zeros_like(ximg_fit)
+    normfit[imed] = normimg
+    spat_set, outmask_spat, spatfit, _ = utils.bspline_profile(ximg_fit, normfit, imed.astype(float),np.ones_like(normfit),
+                                                               inmask = imed, nord=4,upper=5.0, lower=5.0,fullbkpt = fullbkpt)
 
+    # Evaluate and save
+    illumflat = np.ones_like(flat)
+    illumflat[thismask], _ = spat_set.value(ximg[thismask])
+    norm_spec_spat = np.ones_like(flat)
+    norm_spec_spat[thismask] = flat[thismask]/np.fmax(spec_model[thismask], 1.0)/np.fmax(illumflat[thismask],0.01)
+
+    if tweak_slits:
+        tweak_left = False
```

Comment:
Pull this tweaking algorithm into a function.  How robust is this algorithm?  I.e., do you have a good sense of when/if it will break?

Reply:
I've moved this algorithm into a function as you requested.